cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -117,27 +117,42 @@ Status AzureOptions::ConfigureClientSecretCredential(const std::string& account_
                                                      const std::string& tenant_id,
                                                      const std::string& client_id,
                                                      const std::string& client_secret) {
+  account_name_ = account_name;
   credential_kind_ = CredentialKind::kTokenCredential;
   token_credential_ = std::make_shared<Azure::Identity::ClientSecretCredential>(
       tenant_id, client_id, client_secret);
   return Status::OK();
 }
 
 Status AzureOptions::ConfigureDefaultCredential(const std::string& account_name) {
+  account_name_ = account_name;
   credential_kind_ = CredentialKind::kTokenCredential;
   token_credential_ = std::make_shared<Azure::Identity::DefaultAzureCredential>();
   return Status::OK();
 }
 
+Status AzureOptions::ConfigureManagedIdentityCredential(const std::string& account_name,
+                                                        const std::string& client_id) {
+  account_name_ = account_name;
+  credential_kind_ = CredentialKind::kTokenCredential;
+  token_credential_ =
+      std::make_shared<Azure::Identity::ManagedIdentityCredential>(client_id);
+  return Status::OK();
+}
+
 Status AzureOptions::ConfigureWorkloadIdentityCredential(
     const std::string& account_name) {
+  account_name_ = account_name;
   credential_kind_ = CredentialKind::kTokenCredential;
   token_credential_ = std::make_shared<Azure::Identity::WorkloadIdentityCredential>();
   return Status::OK();
 }
 
 Result<std::unique_ptr<Blobs::BlobServiceClient>> AzureOptions::MakeBlobServiceClient()
     const {
+  if (account_name_.empty()) {
+    return Status::Invalid("AzureOptions doesn't contain a valid account name");
+  }
```

Comment:
Another simplification we can make:

 - Rename `kAnonymous` to `kDefaultCredential`
 - On the `kDefaultCredential` case here, initialize `token_credential_` to `std::make_shared<Azure::Identity::DefaultAzureCredential>()` if it's not initialized already.
 
 In the `kTokenCredential` case you can add a `DCHECK(token_credential_)` as documentation that when `credential_kind_` is set to `kTokenCredential` we expect it to be set already (something that is guaranteed by `credential_kind_` being private and set only by functions that also initialize `token_credential_`.
 
 What you think?

Reply:
I think its useful to support both anonymous and default credential as distinct things. Admittedly, its a niche use-case but storage accounts can be public, which is when anonymous is useful https://learn.microsoft.com/en-us/azure/storage/blobs/anonymous-read-access-configure?tabs=portal.

Its a good question though which should be the default. I know `adlfs` deafults to anonomous rather than default credential. In some respects it might be good to be consistent but also this choice of default has caused issues for me in the past e.g. https://github.com/flyteorg/flyte/issues/3962

Reply:
Ok. Makes sense to keep `kAnonymous` then. But what would you return in `AzureOptions::MakeBlobServiceClient()` when `credential_kind_` is anonymous? Lazy-init `token_credential_` with `DefaultAzureCredential` and create the client with it?

Reply:
I looked and there is a constructor that doesn't take any credentials. I guess that's what we should call :-)

Reply:
:+1: Do you mind if we do that in a separate PR to fix up anonymous and change the default to default credential.

Reply:
No problem.